src/Raven.Server/Documents/Indexes/MapReduce/AutoMapReduceIndex.cs

```diff
@@ -197,66 +187,81 @@ public override IQueryResultRetriever GetQueryResultRetriever(DocumentsOperation
             return new MapReduceQueryResultRetriever(indexContext, fieldsToFetch);
         }
 
-        private Table GetMapEntriesTable(Transaction tx)
+        private static Tree GetMapEntriesTree(Transaction tx)
         {
-            _mapEntriesSchema.Create(tx, "MapResults");
-            var table = new Table(_mapEntriesSchema, "MapResults", tx);
+            // map entries structure
+            // MapEntries tree has the following entries
+            // -> { document key, fixed size tree }
+            // each fixed size tree stored records like 
+            // -> { identifier of a map result, hash of a reduce key for the map result }
 
-            return table;
+            return tx.CreateTree("MapEntries");
         }
 
         private unsafe void PutMappedResult(BlittableJsonReaderObject mappedResult, LazyStringValue documentKey, ulong reduceKeyHash, ReduceKeyState state, TransactionOperationContext indexContext)
         {
-            var tvb = new TableValueBuilder
-            {
-                { documentKey.Buffer, documentKey.Size },
-                { (byte*) &reduceKeyHash, sizeof(ulong) },
-                { mappedResult.BasePointer, mappedResult.Size }
-            };
+            var documentMapEntries = _mapReduceWorkContext.MapEntries.FixedTreeFor(documentKey, sizeof(ulong));
 
-            var existingEntry = GetMapEntryForDocument(documentKey);
+            long id = -1;
 
-            long storageId;
-            if (existingEntry == null)
-            {
-                storageId = _mapReduceWorkContext.MapEntriesTable.Insert(tvb);
-            }
-            else
+            if (documentMapEntries.NumberOfEntries > 0)
             {
-                if (existingEntry.ReduceKeyHash != reduceKeyHash)
+                var mapEntries = GetMapEntriesForDocument(documentKey, documentMapEntries);
+
+                if (mapEntries.Count == 1 && mapEntries[0].ReduceKeyHash == reduceKeyHash)
+                {
+                    // update of existing entry, reduce key remained the same - we are going to overwrite the map result only
+                    id = mapEntries[0].Id;
+                }
+                else
                 {
-                    // reduce key changed - need to remove record in related reduce tree
+                    foreach (var mapEntry in mapEntries)
+                    {
+                        var previousState = GetReduceKeyState(mapEntry.ReduceKeyHash, indexContext, create: false);
+
+                        fixed (long* ptr = &mapEntry.Id)
```

Comment:
copying to a  local argument (whose address we can take directly), can likely reduce the amount of (`fixed`) costs significantly here.